src/test/java/com/wikia/webdriver/common/core/url/UrlBuilder.java

```diff
@@ -50,115 +44,130 @@ public String getUrlForPage(String wikiName, String pageName) {
   }
 
   public String getUrlForPage(Page page, String qs) {
     return appendQueryStringToURL(page.getUrl(), qs);
   }
 
+  public String getUrlForPath(String wikiName, String language, String wikiPath) {
+    return addPathToUrl(getUrlForWiki(wikiName, language, false, envType), wikiPath);
+  }
+
   public String getUrlForPath(String wikiName, String wikiPath) {
     return addPathToUrl(getUrlForWiki(wikiName), wikiPath);
   }
 
-  public String getUrlForPath(String wikiName, String wikiPath, EnvType envType) {
-    return addPathToUrl(getUrlForWiki(wikiName, envType), wikiPath);
+  public String getUrlForPath(String wikiPath) {
+    return getUrlForPath(Configuration.getWikiName(), Configuration.getLanguage(), wikiPath);
   }
 
   protected String addPathToUrl(String url, String path) {
-
-    String outputUrl = (!path.startsWith("/")) ? String.format("%s/%s", url, path): String.format("%s%s", url, path);
+    HttpUrl.Builder urlBuilder = HttpUrl.parse(url).newBuilder();
+    urlBuilder.addEncodedPathSegments(path).build();
 
     String qs = Configuration.getQS();
     if (StringUtils.isNotBlank(qs)) {
-      outputUrl = appendQueryStringToURL(outputUrl, qs);
+      urlBuilder.encodedQuery(qs);
     }
-    return outputUrl;
+    return urlBuilder.build().toString();
   }
 
-  public String getUrlForPath(String wikiPath) {
-    return getUrlForPath(Configuration.getWikiName(), wikiPath);
+  public String getUrlForWiki() {
+    return getUrlForWiki(Configuration.getWikiName(), Configuration.getLanguage(), false);
   }
 
-  public String getUrlForWiki() {
-    return getUrlForWiki(Configuration.getWikiName(), false);
+  public String getUrlForWiki(String wikiName, String language, boolean addWWW) {
+    return getUrlForWiki(wikiName, language, addWWW, envType);
   }
 
   public String getUrlForWiki(String wikiName) {
     return getUrlForWiki(wikiName, false);
   }
 
-  public String getUrlForWiki(String wikiName, EnvType envType) {
-    return getUrlForWiki(wikiName, false, envType);
-  }
-
-  public String getUrlForWiki(boolean addWWW) {
-    return getUrlForWiki(Configuration.getWikiName(), addWWW);
-  }
-
   public String getUrlForWiki(String wikiName, boolean addWWW) {
     return getUrlForWiki(wikiName, addWWW, envType);
   }
 
 
   public String getUrlForWiki(String wikiName, boolean addWWW, EnvType envType) {
+    return getUrlForWiki(wikiName, DEFAULT_LANGUAGE, addWWW, envType);
+  }
+
+  public String getUrlForWiki(String wikiName, String language) {
+    return getUrlForWiki(wikiName, language, false, envType);
+  }
+
+  public String getUrlForWiki(String wikiName, String language, boolean addWWW, EnvType envType) {
     final String wikiaName = getWikiaGlobalName(wikiName);
 
     String www = "";
     if (addWWW) {
       www = "www.";
     }
 
+    String host;
+    HttpUrl.Builder urlBuilder = new HttpUrl.Builder();
     switch (envType) {
       case DEV: {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Extract this nested code block into a method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar-s4/coding_rules#rule_key=squid%3AS1199)